node/src/components/console/tasks.rs

```diff
@@ -0,0 +1,335 @@
+    async fn send_to_client<T>(&self, writer: &mut OwnedWriteHalf, response: &T) -> io::Result<()>
+    where
+        T: Display + Serialize,
+    {
+        match self.output {
+            OutputFormat::Interactive => {
+                writer.write_all(response.to_string().as_bytes()).await?;
+                writer.write_all(b"\n").await?;
+            }
+            OutputFormat::Json => {
+                let buf = serde_json::to_string_pretty(response)
+                    .map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+                writer.write_all(buf.as_bytes()).await?;
+                writer.write_all(b"\n").await?;
+            }
+            OutputFormat::Bincode => {
+                let buf = bincode::serialize(response)
+                    .map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+                writer.write_all(&buf).await?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Handler for client connection.
+///
+/// The core loop for the console; reads commands via unix socket and processes them accordingly.
+///
+/// # Security
+///
+/// The handler itself will buffer an unlimited amount of data if no newline is encountered in the
+/// input stream. For this reason ensure that only trusted client connect to the socket producing
+/// the passed in `stream`.
+async fn handler<REv>(
+    effect_builder: EffectBuilder<REv>,
+    stream: UnixStream,
+    mut shutdown_receiver: watch::Receiver<()>,
+) -> io::Result<()>
+where
+    REv: From<DumpConsensusStateRequest> + Send,
+{
+    debug!("accepted new connection on console socket");
+
+    let (reader, mut writer) = stream.into_split();
+    let mut lines = BufReader::new(reader).lines();
+    let mut session = Session::default();
+
+    loop {
+        let shutdown_messages = async { while shutdown_receiver.changed().await.is_ok() {} };
+
+        match future::select(Box::pin(shutdown_messages), Box::pin(lines.next_line())).await {
+            Either::Left(_) => {
+                info!("shutting down console connection to client");
+                break Ok(());
+            }
+            Either::Right((line_result, _)) => {
+                if let Some(line) = line_result? {
+                    session
+                        .process_line(effect_builder, &mut writer, line.as_str())
+                        .await?;
+                } else {
+                    info!("client closed console connection");
+                    break Ok(());
+                }
+            }
+        }
+    }
+}
+
+/// Server task for console.
+pub(super) async fn server<REv>(
+    effect_builder: EffectBuilder<REv>,
+    socket_path: PathBuf,
+    listener: UnixListener,
+    mut shutdown_receiver: watch::Receiver<()>,
+) where
+    REv: From<DumpConsensusStateRequest> + Send,
+{
+    let handling_shutdown_receiver = shutdown_receiver.clone();
+    let mut next_client_id: u64 = 0;
```

Comment:
Shouldn't it be `AtomicU32` or sth?

Reply:
There's no one else accessing this, so no atomic necessary!